src/super_gradients/common/environment/ddp_utils.py

```diff
@@ -77,3 +82,159 @@ def find_free_port() -> int:
+    return list(range(get_world_size()))
+
+
+def count_used_devices() -> int:
+    return len(get_device_ids())
+
+
+def execute_and_distribute_from_master(func: Callable[..., Any]) -> Callable[..., Any]:
+    """
+    Decorator to execute a function on the master process and distribute the result to all other processes.
+    Useful in parallel computing scenarios where a computational task needs to be performed only on the master
+    node (e.g., a computational-heavy calculation), and the result must be shared with other nodes without
+    redundant computation.
+
+    Example usage:
+        >>> @execute_and_distribute_from_master
+        >>> def some_code_to_run(param1, param2):
+        >>>     return param1 + param2
+
+    The wrapped function will only be executed on the master node, and the result will be propagated to all
+    other nodes.
+
+    :param func:    The function to be executed on the master process and whose result is to be distributed.
+    :return:        A wrapper function that encapsulates the execute-and-distribute logic.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        # Run the function only if it's the master process
+        if device_config.assigned_rank <= 0:
+            result = func(*args, **kwargs)
+        else:
+            result = None
+
+        # Broadcast the result from the master process to all nodes
+        return broadcast_from_master(result)
+
+    return wrapper
+
+
+def broadcast_from_master(data: Any) -> Any:
+    """
+    Broadcast data from master node to all other nodes. This may be required when you
+    want to compute something only on master node (e.g computational-heavy metric) and
+    don't want to vaste CPU of other nodes doing same work simultaneously.
+
+    >>> if device_config.assigned_rank <= 0:
+    >>>    result = some_code_to_run(...)
+    >>> else:
+    >>>    result = None
+    >>> # 'result' propagated to all nodes from master
+    >>> result = broadcast_from_master(result)
+
+    :param data:    Data to be broadcasted from master node (rank 0)
+    :return:        Data from rank 0 node
+    """
+    world_size = get_world_size()
+    if world_size == 1:
+        return data
+
+    local_rank = get_local_rank()
+    storage: torch.Tensor
+
+    if local_rank == 0:
+        buffer = pickle.dumps(data)
+        storage = torch.ByteStorage.from_buffer(buffer)
+        payload = torch.ByteTensor(storage).to("cuda")
+        local_size = payload.numel()
+    else:
+        local_size = 0
+
+    # Propagate target tensor size to all nodes
+    local_size = max(all_gather(local_size))
+    if local_rank != 0:
+        payload = torch.empty((local_size,), dtype=torch.uint8, device="cuda")
+
+    dist.broadcast(payload, 0)
+    buffer = payload.cpu().numpy().tobytes()
+    return pickle.loads(buffer)
+
+
+def all_gather(data: Any) -> List[Any]:
+    """
+    Run all_gather on arbitrary picklable data (not necessarily tensors)
```

Comment:
Please see torch.distributed.all_gather_object.
It can simplify alot of the code here.

Reply:
done [36abc20](https://github.com/Deci-AI/super-gradients/pull/1415/commits/36abc20222d1cc94e0bd50c0b503037e83af1f05)